Add core matcher 'toBeInstanceOf'

elliot-nelson opened the pull request:
## Description
Provide a matcher that gives developers a single way to express "expect `value` to be of type `type`".  The intent is that it can be used for both `typeof` and `instanceof` checks, like so:

Existing Check | With `toBeInstanceOf`
--- | ---
`expect(typeof result).toBe('number')` | `expect(result).toBeInstanceOf(Number)`
`expect(typeof result).toBe('string')` | `expect(result).toBeInstanceOf(String)`
`expect(result instanceof Error).toBe(true)` | `expect(result).toBeInstanceOf(Error)`
`expect(result instanceof MyClass).toBe(true)` | `expect(result).toBeInstanceOf(MyClass)`

## Motivation and Context
Primarily the goal is better failure messages - a slightly more accurate message for `typeof` (primitive) checks, and a dramatically improved message for `instanceof` checks.

```
// Old Messages
Failed: Expected 'number' to be 'string'
Failed: Expected false to be true

// New Messages
Failed: Expected instance of Number to be an instance of String
Failed: Expected instance of Object to be an instance of MyClass
```

## How Has This Been Tested?
- New unit tests (mostly to document the expected behavior)
- Local passes:
  - node v6, v8, v10
  - `JASMINE_BROWSER=firefox node ci.js`
  - `JASMINE_BROWSER=chrome node ci.js` (except for unit test mentioned in https://github.com/jasmine/jasmine/issues/1695)
  - `JASMINE_BROWSER=safari node ci.js`

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

slackersoft commented:
I think you should also be able to use `jasmine.any` to accomplish what you're looking for with Jasmine as it stands now. So:

Existing Check | With jasmine.any
-- | --
expect(typeof result).toBe('number') | expect(result).toEqual(jasmine.any(Number))
expect(typeof result).toBe('string') | expect(result).toEqual(jasmine.any(String))
expect(result instanceof Error).toBe(true) | expect(result).toEqual(jasmine.any(Error))
expect(result instanceof MyClass).toBe(true) | expect(result).toEqual(jasmine.any(MyClass))

The assertions certainly don't read quite as nicely as an actual `toBeInstanceOf` check, but they should provide you with a similar level of error message when they fail.

I still think a `toBeInstanceOf` matcher could be useful in giving _better_ messages, but it might make sense to take a quick look at `jasmine.any` and see what kind of shared detection logic we can pull out to ensure they match the same things.

elliot-nelson commented:
:+1: Great point!

- Modified matcher source to make use of existing jasmine functionality, rather than doing it myself.
- Updated a few minor specs to reflect how this changes the behavior.  (For the better, mostly, since it works exactly like it would if you used `jasmine.any()` matchers.)

Possible additional changes:

- Test bed might be overkill if we're relying on `jasmine.any()`.  I could strip it down to a few basic examples showing we're passing through to `Any`, plus the corner cases for null/undefined/etc.  On the flip side, keeping the test bed would notify anyone who makes substantial future changes to `jasmine.Any` that they are making potential breaking changes to the behavior of this matcher...

- Naming... `toBeInstanceOf(Number)` feels a little better and even more conversational as `.toBeA(String)`, `.toBeA(Number)`.  But, you'd probably need to support the _alias_ `.toBeAn(Object)`, `.toBeAn(Error)` as while (not sure if we have any precedent for that yet).  Distant third would be `.toBeOfType(Number)` -- clear, but pretty formal sounding.

slackersoft commented:
I think this looks good. I like the fully specified tests here to clearly state how it works. I think the current name is the clearest.

I'm not entirely sure what's going on with the PhantomJS build for this, but you might be able to rebase onto master to get the CI fixes and see if that gets us some more logging.

elliot-nelson commented:
The PhantomJS build correctly caught a raw `async` function.

I've fixed this PR, and I put up a separate PR to improve the error reporting in that situation.
